gtsam/sfm/DsfTrackGenerator.h

```diff
@@ -0,0 +1,207 @@
+  SfmTrack2d(std::vector<NamedSfmMeasurement> &measurements) : measurements_(measurements) {}
+
+  // Add a measurement to the track.
+  void addMeasurement(const NamedSfmMeasurement &m) {
+    measurements_.emplace_back(m);
+  }
+
+  /// The measurement at index `idx`
+  NamedSfmMeasurement measurement(size_t idx) const { return measurements_[idx]; }
+
+  // Return all measurements in the track.
+  std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
+
+  /// Total number of measurements in this track.
+  size_t numberMeasurements() const { return measurements_.size(); }
+
+  // @brief Validates the track by checking that no two measurements are from the same camera.
+  //
+  // returns boolean result of the validation.
+  bool hasUniqueCameras()
+  {
+    std::vector<int> track_cam_idxs;
+    for (auto & measurement : measurements_)
+    {
+      track_cam_idxs.emplace_back(measurement.i);
+    }
+    auto i = std::adjacent_find(track_cam_idxs.begin(), track_cam_idxs.end());
+    bool all_cameras_unique = (i == track_cam_idxs.end());
+    return all_cameras_unique;
+  }
+};
+
+using SfmTrack2dVector = std::vector<gtsam::SfmTrack2d>;
+using NamedSfmMeasurementVector = std::vector<NamedSfmMeasurement>;
+
+
+/**
+ * @brief Generates point tracks from connected components in the keypoint matches graph.
+ */
+class DsfTrackGenerator {
+
+ public:
+  /** Default constructor. */
+  DsfTrackGenerator() {}
+
+  // Creates a list of tracks from 2d point correspondences.
+  //
+  // Creates a disjoint-set forest (DSF) and 2d tracks from pairwise matches.
+  // We create a singleton for union-find set elements from camera index of a
+  // detection and the index of that detection in that camera's keypoint list,
+  // i.e. (i,k).
+  // @param Map from (i1,i2) image pair indices to (K,2) matrix, for K
+  //        correspondence indices, from each image.
+  // @param Length-N list of keypoints, for N images/cameras.
+  std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+    const MatchIndicesMap& matches_dict,
+    const KeypointsList& keypoints_list) {
+    std::vector<SfmTrack2d> track_2d_list;
+
+    std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
+    // Generate the DSF to form tracks.
+    DSFMapIndexPair dsf;
+
+    for (const auto& kv : matches_dict) {
+      const auto pair_idxs = kv.first;
+      const auto corr_idxs = kv.second;
+
+      // Image pair is (i1,i2).
+      size_t i1 = pair_idxs.first;
+      size_t i2 =  pair_idxs.second;
+      for (size_t k = 0; k < corr_idxs.rows(); k++)
+      {
+        // Measurement indices are found in a single matrix row, as (k1,k2).
+        size_t k1 = corr_idxs(k, 0);
+        size_t k2 = corr_idxs(k, 1);
+        // Unique keys for the DSF are (i,k), representing keypoint index in an image.
+        dsf.merge(IndexPair(i1, k1), IndexPair(i2, k2));
+      }
+    }
+
+    std::cout << "[SfmTrack2d] Union-Find Complete" << std::endl;
```

Comment:
do we need this?

Reply:
Thanks, added a `verbose` flag to the method that defaults to false that controls the logging verbosity.

For very large sequences, this logging can provide helpful visibility during extended execution regarding the current execution state.